### ISSUE TITLE
Disable monitoring not working?

### ISSUE THREAD

**Issue by ctindel** [elastic/logstash-docker#65]
I use       XPACK_MONITORING_ENABLED: "false" in my docker-compose file.  

I can see when I peek into my container that it is set:

```
# env | grep -i MONI
XPACK_MONITORING_ENABLED=false
```

However that doesn't appear to be changing the logstash.yml file the way the docs page (https://www.elastic.co/guide/en/logstash/current/_configuring_logstash_for_docker.html) says it will:

```
# cat logstash.yml
http.host: "0.0.0.0"
path.config: /usr/share/logstash/pipeline
xpack.monitoring.elasticsearch.url: http://elasticsearch:9200
xpack.monitoring.elasticsearch.username: logstash_system
xpack.monitoring.elasticsearch.password: changeme
```

So I'm left with tons of monitoring related error messages:

```
sa-demo-apachelogs-ls-build        | [2017-11-22T17:54:36,204][INFO ][logstash.inputs.metrics  ] Monitoring License OK
sa-demo-apachelogs-ls-build        | [2017-11-22T17:54:36,300][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://elasticsearch:9200/_xpack/monitoring/_bulk?system_id=logstash&system_api_version=2&interval=1s"}
sa-demo-apachelogs-ls-build        | [2017-11-22T17:54:38,311][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://elasticsearch:9200/_xpack/monitoring/_bulk?system_id=logstash&system_api_version=2&interval=1s"}
sa-demo-apachelogs-ls-build        | [2017-11-22T17:54:42,393][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://elasticsearch:9200/_xpack/monitoring/_bulk?system_id=logstash&system_api_version=2&interval=1s"}
sa-demo-apachelogs-ls-build        | [2017-11-22T17:54:50,408][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://elasticsearch:9200/_xpack/monitoring/_bulk?system_id=logstash&system_api_version=2&interval=1s"}
sa-demo-apachelogs-ls-build        | [2017-11-22T17:55:06,415][ERROR][logstash.outputs.elasticsearch] Encountered a retryable error. Will Retry with exponential backoff  {:code=>400, :url=>"http://elasticsearch:9200/_xpack/monitoring/_bulk?system_id=logstash&system_api_version=2&interval=1s"}
```

**Comment by ctindel** [comment]
It also prevents logstash from shutting down properly:

```
sa-demo-apachelogs-ls-build        | [2017-11-22T17:57:44,537][WARN ][logstash.shutdownwatcher ] {"inflight_count"=>1, "stalling_thread_info"=>{"other"=>[{"thread_id"=>27, "name"=>"[.monitoring-logstash]>worker0", "current_call"=>"[...]/vendor/bundle/jruby/2.3.0/gems/stud-0.0.23/lib/stud/interval.rb:89:in `sleep'"}]}}
```

**Comment by jarpy** [comment]
Please specify the version whenever raising issues. It helps to replicate problems. This one can't be replicated on version 6.0.0, for example.

```
$ docker run -e XPACK_MONITORING_ENABLED=false docker.elastic.co/logstash/logstash:6.0.0 cat /usr/share/logstash/config/logstash.yml
2017/11/23 00:56:26 Setting 'xpack.monitoring.enabled' from environment.
http.host: 0.0.0.0
path.config: /usr/share/logstash/pipeline
xpack.monitoring.elasticsearch.password: changeme
xpack.monitoring.elasticsearch.url: http://elasticsearch:9200
xpack.monitoring.elasticsearch.username: logstash_system
xpack.monitoring.enabled: false
```

It's also good to share what you did to produce the error. It's _really_ hard to get the same failure as somebody else when you don't know what they did. It's likely that something in your approach is preventing the environment-to-yaml mapping step from happening, but I can't tell without seeing the reproduction case.

This setting is also covered by an acceptance test, so it's unlikely (though never impossible) that it's a bug.

https://github.com/elastic/logstash-docker/blob/master/tests/test_settings.py#L35,L37

If it's not a bug, then it doesn't need a GitHub issue.

Finally, if you don't want free Monitoring (and assuming you are using version 6), I suggest using the open source image.

**Comment by ctindel** [comment]
@jarpy sorry, you're right I was in a hurry because of the holiday and made a sloppy report.  It was with 6.0.  I build a new logstash image using 6.0.0 as the base.  I am pretty sure this is how I did it in 5.0 as well and using the environment variable worked.  Though I guess it's possible that maybe it didn't work and some behavior changed in logstash in 6.0 that breaks it?

If you do a "docker-compose up" to build and start the image, then do:

```
# docker exec -it logstash-test-build cat /usr/share/logstash/config/logstash.yml
http.host: "0.0.0.0"
path.config: /usr/share/logstash/pipeline
xpack.monitoring.elasticsearch.url: http://elasticsearch:9200
xpack.monitoring.elasticsearch.username: logstash_system
xpack.monitoring.elasticsearch.password: changeme
```

Tar file attached.
[docker_ls_test.tar.gz](https://github.com/elastic/logstash-docker/files/1499807/docker_ls_test.tar.gz)

**Comment by jarpy** [comment]
Thanks.

Your `Dockerfile` is overriding the `ENTRYPOINT`, which is [where all the magic that you want is implemented](https://github.com/elastic/logstash-docker/blob/master/build/logstash/bin/docker-entrypoint#L3,L7).

**Comment by ctindel** [comment]
Hmmm ok. So if you run your own alternate command you have to have your entry point script call that magic script? Is that a level of api we want to support so they can get the magic?

**Comment by ctindel** [comment]
It's a nuance that has always escaped me.

So if my docker compose just sets the command instead of having a new entrypoint it would work ? Thanks I'll try that.

**Comment by jarpy** [comment]
It's not highly intuitive, I agree. `ENTRYPOINT` _is_ confusing, which is why I don't recommend it unless you can precisely define why you need it.

This image uses what is sometimes called a "smart" entrypoint. [If you ask for something in particular with `CMD`, you'll get it, but if you don't ask for anything, or your `CMD` looks like Logstash flags, you get Logstash](
https://github.com/elastic/logstash-docker/blob/74f123eeb0cc32e33256c9e8ebdd9682a5b4d4f7/build/logstash/bin/docker-entrypoint#L9,L13).

So, in summary, _pay no attention to the script behind the curtain and set `CMD` to whatever you want to do._
